File: src/gestorAplicacion/usuario/Usuario.java
package gestorAplicacion.usuario;

import java.io.Serializable;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import gestorAplicacion.economia.*;


public class Usuario implements Serializable {

    private static final long serialVersionUID = -64431385135968757L;
    
    //Datos necesarios para la identificación del usuario
    private String cedula;
    private String nombre;
    private String email;
    private LocalDate fechaIngreso;
    private String clave;
    //Atributos necesarios para la interacción del usuario con el sistema
    private List<Bolsillo> bolsillos = new ArrayList<>();
    private List<Colchon> colchones = new ArrayList<>();
    private List<Ingreso> ingresos = new ArrayList<>();
    private List<Salida> salidas = new ArrayList<>();
    private List<Prestamo> prestamos = new ArrayList<>();
    private List<Meta> metas = new ArrayList<>();


    public Usuario(String cedula, String nombre, String email, LocalDate fechaIngreso, String clave) {
        setCedula(cedula);
        setNombre(nombre);
        setEmail(email);
        setFechaIngreso(fechaIngreso);
        setClave(clave);
        bolsillos.add(new Bolsillo(this, Divisa.COP, "DEFAULT"));
    }

    public String getCedula() {
        return cedula;
    }

    public void setCedula(String cedula) {
        this.cedula = cedula;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public LocalDate getFechaIngreso() {
        return fechaIngreso;
    }

    public void setFechaIngreso(LocalDate fechaIngreso) {
        this.fechaIngreso = fechaIngreso;
    }

    public String getClave() {
        return clave;
    }

    public void setClave(String clave) {
        this.clave = clave;
    }

    public List<Bolsillo> getBolsillos() {
        return bolsillos;
    }

    public void setBolsillos(List<Bolsillo> bolsillos) {
        this.bolsillos = bolsillos;
    }

    public List<Colchon> getColchones() {
        return colchones;
    }

    public void setColchones(List<Colchon> colchones) {
        this.colchones = colchones;
    }

    public List<Ingreso> getIngresos() {
        return ingresos;
    }

    public void setIngresos(List<Ingreso> ingresos) {
        this.ingresos = ingresos;
    }

    public List<Salida> getSalidas() {
        return salidas;
    }

    public void setSalidas(List<Salida> salidas) {
        this.salidas = salidas;
    }

    public List<Prestamo> getPrestamos() {
        return prestamos;
    }

    public void setPrestamos(List<Prestamo> prestamos) {
        this.prestamos = prestamos;
    }

    public List<Meta> getMetas() {
        return metas;
    }

    public void setMetas(List<Meta> metas) {
        this.metas = metas;
    }
    
    //Se realiza un deposito en la cuenta destino del usuario y se genera un ingreso en el historial
    public void nuevoIngreso(Ingreso ingreso) {
        ingreso.getCuentaDestino().depositar(ingreso.getValorDestino());
        ingresos.add(ingreso);
    }
    
  //Se realiza un retiro validando su consistencia origen del usuario y se genera una salida en el historial
    public boolean nuevaSalida(Salida salida) {
        boolean retirado = salida.getCuentaOrigen().retirar(salida.getValorOrigen());
        if(retirado){
            salidas.add(salida);
        }
        return retirado;
    }

    public void nuevoBolsillo(Bolsillo bolsillo) {
        bolsillos.add(bolsillo);
    }

    public void nuevoColchon(Colchon colchon) {
        colchones.add(colchon);
    }

    public void nuevaMeta(Meta meta) {
        metas.add(meta);
    }

    public void nuevoPrestamo(PrestamoLargoPlazo prestamo, Bolsillo bolsillo) {
        prestamos.add(prestamo);
        bolsillo.depositar(prestamo.getDivisa().ConvertToDivisa(prestamo.getValorInicial(), bolsillo.getDivisa())[0]);
    }
    public void nuevoPrestamo(PrestamoFugaz prestamo, Bolsillo bolsillo) {
        prestamos.add(prestamo);
        bolsillo.depositar(prestamo.getDivisa().ConvertToDivisa(prestamo.getValorInicial(), bolsillo.getDivisa())[0]);
    }
    
    //Se realiza una separacion del dinero del usuario por divisas guardada en bolsillos, colchones y metas
    public double[] getDineroTotal() {
        double[] total = new double[Divisa.values().length];
        List<Contable> contables = new ArrayList<>();
        contables.addAll(getBolsillos());
        contables.addAll(getColchones());
        contables.addAll(getMetas());
        for (Contable i : contables) {
            for (int j = 0; j < Divisa.values().length; j++) {
                if (i.getDivisa().equals(Divisa.values()[j])) {
                    total[j] += i.getSaldo();
                    break;/**
     *
     */
                }
            }
        }
        return total;
    }
 
}

File: Python/gestorAplicacion/usuario/Usuario.py
from gestorAplicacion.usuario.Bolsillo import Bolsillo

class Usuario:

    def __init__(self, cedula, nombre, email, fechaIngreso, clave):
        #Datos necesarios para la identificación del usuario
        self._cedula = cedula
        self._nombre = nombre
        self._email = email
        self._fechaIngreso = fechaIngreso
        self._clave = clave
        #Atributos necesarios para la interacción del usuario con el sistema
        self._bolsillos = []
        self._colchones = []
        self._ingresos = []
        self._salidas = []
        self._prestamos = []
        self._metas = []
        self._bolsillos.append(Bolsillo(self, "COP", "DEFAULT"))

    def getCedula(self):
        return self._cedula

    def setCedula(self, cedula):
        self._cedula = cedula

    def getNombre(self):
        return self._nombre

    def setNombre(self, nombre):
        self._nombre = nombre

    def getEmail(self):
        return self._email

    def setEmail(self, email):
        self._email = email

    def getFechaIngreso(self):
        return self._fechaIngreso

    def setFechaIngreso(self, fechaIngreso):
        self._fechaIngreso = fechaIngreso

    def getClave(self):
        return self._clave

    def setClave(self, clave):
        self._clave = clave

    def getBolsillos(self):
        return self._bolsillos

    def setBolsillos(self, bolsillos):
        self._bolsillos = bolsillos

    def getColchones(self):
        return self._colchones

    def setColchones(self, colchones):
        self._colchones = colchones

    def getIngresos(self):
        return self._ingresos

    def setIngresos(self, ingresos):
        self._ingresos = ingresos

    def getSalidas(self):
        return self._salidas

    def setSalidas(self, salidas):
        self._salidas = salidas

    def getPrestamos(self):
        return self._prestamos

    def setPrestamos(self, prestamos):
        self._prestamos = prestamos

    def getMetas(self):
        return self._metas

    def setMetas(self, metas):
        self._metas = metas

    #Se realiza un deposito en la cuenta destino del usuario y se genera un ingreso en el historial
    def nuevoIngreso(self, ingreso):
        ingreso.getCuentaDestino().depositar(ingreso.getValorDestino())
        self._ingresos.append(ingreso)

    #Se realiza un retiro validando su consistencia origen del usuario y se genera una salida en el historial
    def nuevaSalida(self, salida):
        retirado = salida.getCuentaOrigen().retirar(salida.getValorOrigen())
        if retirado:
            self._salidas.append(salida)
        return retirado

    def nuevoBolsillo(self, bolsillo):
        self._bolsillos.append(bolsillo)

    def nuevoColchon(self, colchon):
        self._colchones.append(colchon)

    def nuevaMeta(self, meta):
        self._metas.append(meta)

    def nuevoPrestamo(self, prestamo, bolsillo):
        self._prestamos.append(prestamo)
        bolsillo.depositar(prestamo.getDivisa().ConvertToDivisa(prestamo.getValorInicial(), bolsillo.getDivisa())[0])

    def nuevoPrestamo(self, prestamo, bolsillo):
        self._prestamos.append(prestamo)
        bolsillo.depositar(prestamo.getDivisa().ConvertToDivisa(prestamo.getValorInicial(), bolsillo.getDivisa())[0])


